mlir-sycl/lib/Conversion/SYCLToGPU/SYCLToGPU.cpp

```diff
@@ -0,0 +1,218 @@
+        loc, targetIndexTy, getDimension(rewriter, loc, opName, i)));
+    const auto ptr = createGetOp(rewriter, loc, dimMtTy, res, index,
+                                 argumentTypes, functionName);
+    rewriter.create<memref::StoreOp>(loc, val, ptr, zero);
+  }
+}
+
+/// Converts n-dimensional operations to operations of name \p opName, from the
+/// GPU dialect.
+///
+/// There are three possible cases here:
+/// 1. No argument is passed (see convertToFullObject())
+/// 2. A constant argument is passed (see converWithConstIndex())
+/// 3. A non-constant argument is passed (see convertWithVarIndex())
+void rewrite(StringRef opName, Operation *op, ValueRange operands,
+             ConversionPatternRewriter &rewriter) {
+  switch (op->getNumOperands()) {
+  case 0: {
+    const auto dimensions = getDimensions(op->getResultTypes()[0]);
+    convertToFullObject(rewriter, opName, op, dimensions);
+    break;
+  }
+  case 1: {
+    const auto dimension = operands[0];
+    if (const auto definingOp = dimension.getDefiningOp<arith::ConstantOp>()) {
+      convertWithConstIndex(rewriter, opName, op, definingOp);
+    } else {
+      // TODO: Do not rely in a default number of dimensions; take into
+      // account kernel dimensionality. This should be available in the parent
+      // function.
+      constexpr int64_t defaultDimensions{3};
+      convertWithVarIndex(rewriter, opName, op, defaultDimensions, dimension);
+    }
+    break;
+  }
+  default:
+    llvm_unreachable("Invalid cardinality");
+  }
+}
+
+/// Converts n-dimensional operations of type \tparam OpTy to operations of type
+/// \tparam GPUOpTy.
+///
+/// Work is offloaded to the rewrite function above.
+template <typename OpTy, typename GPUOpTy>
+class GridOpPattern final : public OpConversionPattern<OpTy> {
+public:
+  using OpConversionPattern<OpTy>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(OpTy op, typename OpTy::Adaptor opAdaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    ::rewrite(GPUOpTy::getOperationName(), op, opAdaptor.getOperands(),
+              rewriter);
+    return success();
+  }
+};
+
+/// Converts one-dimensional operations of type \tparam OpTy to operations of
+/// type \tparam GPUOpTy.
+///
+/// Due to the different output type, as casting is needed before returning.
+template <typename OpTy, typename GPUOpTy>
+class SingleDimGridOpPattern final : public OpConversionPattern<OpTy> {
+public:
+  using OpConversionPattern<OpTy>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(OpTy op, typename OpTy::Adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    // Create the new operation
+    const auto res = static_cast<Value>(rewriter.create<GPUOpTy>(op->getLoc()));
+    // And cast
+    rewriter.replaceOpWithNewOp<arith::IndexCastOp>(op, op->getResultTypes()[0],
+                                                    res);
+    return success();
+  }
+};
```

Comment:
For "1-D" operations, we just create a new operation and cast to the target type.